### PR TITLE
Implement common console launcher

### LINKED ISSUE #600: Implement console launcher for Kotlin Native
Relates to #555 

For prior art see: https://github.com/spekframework/spek/blob/2.x/spek-ide-plugin/interop-jvm/src/main/kotlin/org/spekframework/ide/console.kt.

I'm imagining something like:

```kotlin
interface ConsoleLauncher {
  fun launch(context: DiscoveryContext, args: List<String>): Int
}
```

then KN user's in the meantime can write:

```kotlin
fun main(args: List<String>) {
   // build discovery context
   val discoveryContext = DiscoveryContext.builder()
      ...
      .build()

  val consoleLauncher = KNConsoleLauncher()
  val exitCode = consoleLauncher.launch(discoveryContext, args)
  exitProcess(exitCode)
}
```

Within gradle, we can invoke something like:

```
<executable> --reporter=xunit --reports-dir={project-build-dir}/spek/reports
```

For IJ integration (eventually), we can issue a command like (`sm` is service message, see https://github.com/spekframework/spek/blob/2.x/spek-ide-plugin/interop-jvm/src/main/kotlin/org/spekframework/ide/ServiceMessageAdapter.kt):
```
<executable> --path <path-to-run> --reporter=console --console-reporter-format=sm
```

The actual reporter implementations can be done separately, this issue only covers the necessary infrastructure/api to write custom reporters.

### PR DESCRIPTION
**Pull request by raniejade**
@charleskorn resolves #600 

I think the infrastructure/api is mostly there, the only bit is parsing the command line arguments and wiring them up.

### CODE REVIEW

File: spek-runtime/src/nativeMain/kotlin/org/spekframework/spek2/launcher/console.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package org.spekframework.spek2.launcher
+
+actual class ConsoleLauncher: AbstractConsoleLauncher() {
+    override fun parseArgs(args: List<String>): LauncherArgs {
+        return LauncherArgs(
+            listOf(ConsoleReporterType(ConsoleReporterType.Format.BASIC)),
```
Comment (raniejade) [inline spek-runtime/src/nativeMain/kotlin/org/spekframework/spek2/launcher/console.kt]: @charleskorn eventually we need to parse `args` here, but for now lets send the default emulating what you had before.

File: spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/launcher/ConsoleLauncher.kt [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+sealed class ReporterType
+class ConsoleReporterType(val format: Format): ReporterType() {
+    enum class Format {
+        BASIC,
+        SERVICE_MESSAGE
+    }
+}
+
+data class LauncherArgs(
+    val reporterTypes: List<ReporterType>,
```
Comment (charleskorn) [inline spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/launcher/ConsoleLauncher.kt:24]: In what scenario would you want to have multiple reporters?

File: integration-test/build.gradle [hunk 3]
```diff
@@ -28,6 +28,7 @@ kotlin {
             dependencies {
                 implementation project(':spek-dsl')
                 implementation kotlin('test')
+                implementation project(':spek-runtime')
```
Comment (charleskorn) [inline integration-test/build.gradle]: Why is this needed?

File: spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/launcher/ConsoleLauncher.kt [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                is ConsoleReporterType -> when (reporter.format) {
+                    ConsoleReporterType.Format.BASIC -> BasicConsoleReporter()
+                    else -> throw AssertionError("Unsupported console reporter: ${reporter.format}")
+                }
+                else -> throw AssertionError("Unsupported reporter: $reporter")
+            }
+        }
+    }
+
+    private fun parseArgs(args: List<String>): LauncherArgs {
```
Comment (raniejade) [inline spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/launcher/ConsoleLauncher.kt:116]: @charleskorn I decided to parse the args in a very naive way.

File: spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/runtime/execution/Discovery.kt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package org.spekframework.spek2.runtime.execution
+
+import org.spekframework.spek2.runtime.discovery.DiscoveryContext
+import org.spekframework.spek2.runtime.scope.Path
+import org.spekframework.spek2.runtime.scope.ScopeImpl
+
+data class DiscoveryRequest(val context: DiscoveryContext, val sourceDirs: List<String>, val paths: List<Path>)
```
Comment (charleskorn) [inline spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/runtime/execution/Discovery.kt:7]: Would it make sense for `sourceDirs` to be on the JVM version of `DiscoveryContext` rather than on `DiscoveryRequest`?

File: spek-runtime/src/nativeMain/kotlin/org/spekframework/spek2/launcher/console.kt (hunk 1 above)
Comment (charleskorn) [reply spek-runtime/src/nativeMain/kotlin/org/spekframework/spek2/launcher/console.kt]: Sounds good to me.

File: spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/launcher/ConsoleLauncher.kt (hunk 2 above)
Comment (raniejade) [reply spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/launcher/ConsoleLauncher.kt:24]: In gradle, sometimes you'd want to generate `xml` and `html` reports at the same time.

File: spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/launcher/ConsoleLauncher.kt [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        var rawReporterType: String? = null
+        var rawConsoleReporterFormat: String? = null
+        var reportExitCode = false
```
Comment (charleskorn) [inline spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/launcher/ConsoleLauncher.kt]: Shouldn't this default to true?

File: spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/launcher/ConsoleLauncher.kt (hunk 2 above)
Comment (charleskorn) [reply spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/launcher/ConsoleLauncher.kt:24]: Right, makes sense.

File: integration-test/build.gradle (hunk 3 above)
Comment (raniejade) [reply integration-test/build.gradle]: removed.

File: spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/runtime/execution/Discovery.kt (hunk 5 above)
Comment (raniejade) [reply spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/runtime/execution/Discovery.kt:7]: Yep, but I'll work on it on a separate PR to minimise the changes here.

File: spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/launcher/ConsoleLauncher.kt (hunk 6 above)
Comment (raniejade) [reply spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/launcher/ConsoleLauncher.kt]: done.

**Comment by charleskorn** [comment]
Looks good to me.

**Review by charleskorn — APPROVED** [review]
